homeassistant/components/smartthings/__init__.py

```diff
@@ -125,6 +125,14 @@
     return all(await asyncio.gather(*tasks))
 
 
+async def async_migrate_entry(hass, entry):
+    """Migrate a previous version config entry."""
+    _LOGGER.debug('Migrating entry from %s to %s', entry.version,
+                  SmartThingsFlowHandler.VERSION)
+    entry.version = SmartThingsFlowHandler.VERSION
```

Comment:
I was just discussing in the options PR #18929 about how we should not allow direct changes to config entries but instead use a method, I think that's the case here too.

So we would want to update the data and the version. Version _could_ be done by `async_migrate`. I think that it would be weird to have to call `hass.config_entries.async_update(…)` inside this method. Hmmm

Reply:
Or maybe we make an exception for migrations? That probably makes the most sense.

CC @Kane610

Reply:
We could do `ConfigEntry.async_migrate` that accepts the fields to update... I would include `title` too.  We'd then except that to be called within the component migration function to make changes.  

The current async_migrate would be made private/rolled into setup, as that's the logic call the migration method on the component.

Reply:
A benefit of that is that it could be called anywhere--not just in the migration handler--for example, if a component knows it needs to upgrade an entry outside of setup.

Reply:
Related PR #20877 . That PR came from the idea that I want to have a central place to update a config entry, so that we can notify the frontend etc.

I have thought more about it, I think it's fine to directly assign to the fields. Let's not make copies and evolve config entries for migrations, we're already changing the entry (changing the state), so this is fine.

Reply:
OK - I'll get the PR cleaned up, remove the ST changes and fix the tests and we can move forward.